### ISSUE TITLE
Missing Usage examples in documentation for visualizations

### ISSUE THREAD

**Issue by modernlearner** [ludwig-ai/ludwig#1067]
It looks like the documentation is still missing code examples of how to use the visualizations: https://ludwig-ai.github.io/ludwig-docs/api/visualization/#compare_performance

It would be nice to see an example of the `test_stats_per_model` parameter:

```
test_stats_per_model (List[dict]): dictionary containing evaluation performance statistics.
```

My code looks like this:

```python
training_stats, preprocessed_data, output_directory = base_model.train(
  training_set='./datasets/spam_train.csv',
  test_set='./datasets/spam_test.csv'
)
```

And this is what I've tried:

```python
visualize.learning_curves(training_stats, ...)
visualize.learning_curves([training_stats], ...)
visualize.learning_curves((training_stats['model'], ...)
```

All of them raised errors. I've seen a blog where the code example supplied a filename string instead of a `List[dict]`, so what's the correct way to call train or evaluate and pass the training statistics to the visualize methods?

Thanks for any help on this!

**Comment by jimthompson5802** [comment]
@modernlearner   I understand the problem you are facing is the inability to generate the Ludwig learning curve visualization.  When you attempt to run `learning_curves()` function you receive error messages.   

First, here is a small program that uses the `learning_curves()` api.  
```
import logging

import pandas as pd
import numpy as np

from sklearn.datasets import make_regression

from ludwig.api import LudwigModel
from ludwig.visualize import learning_curves

# generate synthetic data for training
X, y = make_regression(1000,n_features=3, n_targets=1)
y = y.reshape(-1,1)
df = pd.DataFrame(np.hstack((X, y)))
df.columns=['x1', 'x2', 'x3', 'y']
train_ds, test_ds = train_test_split(df, test_size=0.2)

# setup Ludwig model
input_features = [
    {'name': 'x1', 'type': 'numerical', 'preprocessing': {'normalization': 'zscore'}},
    {'name': 'x2', 'type': 'numerical', 'preprocessing': {'normalization': 'minmax'}},
    {'name': 'x3', 'type': 'numerical'}
]
output_features = [
    {
        'name': 'y', 'type': 'numerical',
        'loss': {'type': 'mean_squared_error'},
        'num_fc_layers': 5, 'fc_size': 64
     }
]

config = {
    'input_features': input_features,
    'output_features': output_features,
    'combiner': {
        'type': 'concat',  # 'concat'
        #'num_fc_layers': 1, 'fc_size': 16
    },
    'training': {
        'epochs': 5,
        'batch_size': 32
    }
}

# train the model
model = LudwigModel(config,logging_level=logging.INFO)
training_statistics, preprocessed_data, output_directory = model.train(training_set=train_ds, test_set=test_ds) 

# generate the learning curve visualization
learning_curves([training_statistics])
```
This is the output from the `learning_curves()` call.
![image](https://user-images.githubusercontent.com/1425269/103041299-05109e80-4544-11eb-9a12-12063426c41f.png)

If this example did not help, then we will need more details on the errors you are seeing.  Please post the exact text of the error messages you are receiving.

**Comment by w4nderlust** [comment]
@modernlearner we recently changed the syntax a bit, so the blogpost you was may not be up to date.
Anyway, I second @jimthompson5802 example, hopefully it helps.
I aslo add that yout
```
visualize.learning_curves([training_stats], ...)
```
should actually be correct, so I would be curious to see the error to try to debug it.

**Comment by modernlearner** [comment]
Thanks for the example @jimthompson5802 

@w4nderlust here's what I'm trying to do:

```python
if base_model_loaded_from_file:
    training_stats, preprocessed_data, output_directory = base_model.evaluate(
        dataset='./datasets/spam_test.csv'
    )
else:
    training_stats, preprocessed_data, output_directory = base_model.train(
        training_set='./datasets/spam_train.csv',
        test_set='./datasets/spam_test.csv',
        skip_save_processed_input=True,
    )
    base_model.save('trained/basic')

visualize.learning_curves(
    [{ 'training': training_stats }],
    None
)
```

So for the `.train` call I would simply use `visualize.learning_curves([training_stats], ...)`. Which does indeed work, thanks again for the example!

How about for the `.evaluate`? Would it be the same thing or would I need to do something else for an already trained model that's been loaded?

**Comment by jimthompson5802** [comment]
@modernlearner   re: `.evaluate`.  This function evaluates predictions derived from a trained model and reports the performance metrics.  Since `.evaluate()` does not do training there is no per epoch `loss` value returned by the `.evaluate()` function.  Hence there is no data for the `learning_curves()` function to display.

The `training_stats` dictionary returned by `.train()` has a different structure than the `eval_stats` dictionary returned in the tuple from `.evaluate()`. 

Please review the description of the return values for [`.train()`](https://ludwig-ai.github.io/ludwig-docs/api/LudwigModel/#train) and [`.evaluate()`](https://ludwig-ai.github.io/ludwig-docs/api/LudwigModel/#evaluate) and note the differences in the description of the contents of the respective "statistics" dictionary contents.

I hope this helps to answer your question.

**Comment by jimthompson5802** [comment]
P.S. In the case of `.evaluate()`, what are you trying to do with the `evaluation_statistics`?  This data structure will contain the performance metrics from evaluation.  One way to see the performance metric is converting the `evaluation_statistics` dictionary into a pandas dataframe and display the contents.  Here is an example
![image](https://user-images.githubusercontent.com/1425269/103063773-53439300-4580-11eb-8aa3-3ae8f84e737e.png)  

You may have to fiddle with the options on the `pandas.DataFrame.from_dict()` to get the statistics in a form that is more useful for your needs.

**Comment by jimthompson5802** [comment]
@modernlearner  In writing about the different content between `training_statistics` and `evaluation_statistics` and showing the content of the difference, I forgot to point out Ludwig's `compare_performance` visualization that works with `evaluation_statistics`.  Here is an example of its use.
![image](https://user-images.githubusercontent.com/1425269/103088796-f49b0b00-45b9-11eb-831e-6a27444cbe60.png)

**Comment by modernlearner** [comment]
@jimthompson5802 I thought `.evaluate` would be the simplest way to get the stats after loading the model from a file instead of re-training (or trying to load the training stats from a JSON file). I did notice that the evaluate stats were missing the epoch data it wasn't clear which other visualizations would work with both training and evaluate stats.

Thanks again for the thorough examples! I created a draft PR to document the use of training and evaluate with compare_performance: https://github.com/ludwig-ai/ludwig/pull/1068

Going to close this issue as my questions have been answered. Thanks again!
